src/Files.App/Actions/FileSystem/FlattenRecursiveAction.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) 2024 Files Community
+// Licensed under the MIT License. See the LICENSE.
+
+using System.IO;
+using Windows.Storage;
+using Microsoft.Extensions.Logging;
+
+namespace Files.App.Actions
+{
+	internal sealed class FlattenRecursiveAction : ObservableObject, IAction
+	{
+		private readonly IContentPageContext context;
+
+		public string Label
+			=> "FlattenRecursive".GetLocalizedResource();
+
+		public string Description
+			=> "FlattenDescription".GetLocalizedResource();
+
+		public bool IsExecutable =>
+			context.ShellPage is not null &&
+			context.HasSelection &&
+			context.SelectedItem?.PrimaryItemAttribute is StorageItemTypes.Folder;
+
+		public FlattenRecursiveAction()
+		{
+			context = Ioc.Default.GetRequiredService<IContentPageContext>();
+
+			context.PropertyChanged += Context_PropertyChanged;
+		}
+
+		public async Task ExecuteAsync(object? parameter = null)
+		{
+			if (context.ShellPage?.ShellViewModel is null)
+				return;
+
+			var items = context.SelectedItems;
+
+			if (items is null || !items.Any() || items.Any(item => !item.IsFolder))
+				return;
+
+			foreach (var item in items)
+			{
+				await FlattenFolderAsync(item.ItemPath);
+			}
+		}
+
+		private async Task FlattenFolderAsync(string folderPath)
+		{
+			var containedFolders = await Task.Run(() => Directory.GetDirectories(folderPath));
+			var containedFiles = await Task.Run(() => Directory.GetFiles(folderPath));
+
+			foreach (var containedFolder in containedFolders)
+			{
+				await FlattenFolderAsync(containedFolder);
```

Comment:
Is it possible to avoid using `await` in the loop?

Reply:
> Is it possible to avoid using `await` in the loop?

Sure, I've pushed a commit changing this to run synchronously.